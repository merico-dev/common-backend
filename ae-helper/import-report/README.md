# import-report

主要负责将 CA 通知来的 protobuf 打包的报告导入到 vdev 的数据库中。

`gen_protos.sh` 脚本用于生成 `protos` 文件夹下面的 protobuf bindings。当 CA 更新了 proto 后，先 `git` 更新一下 `/proto/analysis-proto` 这个 submodule，然后执行这个脚本。

`submit_test_repo.sh` 脚本用于开发阶段提交测试用的分析任务给 CA。

## Notes

* `reportOfAll.fields` 记录了 `reportOfAll` 下哪些字段是有数据的。避免使用 `reportOfAll.meta.analysis_type` 在 vdev 这边硬编码判断。

## TODO

- [ ] 进度通知
