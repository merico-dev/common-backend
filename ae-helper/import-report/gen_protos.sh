mkdir -p ./protos
rm ./protos/all.d.ts
rm ./protos/all.js
npx pbjs -t static-module -w commonjs --root "_1599734255573" --no-create --no-encode --no-verify --no-convert --no-delimited -o ./protos/all.js \
  ../../../proto/analysis-proto/indicator/report_of_all.proto \
  ../../../proto/analysis-proto/indicator/commit.proto \
  ../../../proto/analysis-proto/indicator/line_diff_edit.proto
npx pbts -o ./protos/all.d.ts ./protos/all.js
