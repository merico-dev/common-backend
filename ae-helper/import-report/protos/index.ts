import $root from './all';

import Protos = $root.merico.proto.analysis.indicator;
export default Protos;
