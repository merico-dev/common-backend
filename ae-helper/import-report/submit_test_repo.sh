grpcurl -connect-timeout 30 -max-time 30 -plaintext -d '{
  "git_url": "https://github.com/gliderlabs/ssh.git",
  "project_id": "5024fc90-c18e-43aa-9c78-17b34c08f453",
  "report_id": "1d7fa6f1-151a-4e36-8007-4406d0d80622",
  "path_blacklist": []
}' localhost:9461 merico.analysis.gateway.CodeAnalytics.SubmitRepo
