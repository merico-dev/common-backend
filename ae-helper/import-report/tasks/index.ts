import writeEmails from "./writeEmails";
import writeProjectCommitsTask from "./writeProjectCommitsTask";
import writeReportContributorBombSweeper from './writeReportContributorBombSweeper';
import writeReportCodeCheckBreakingRecordsTask from "./writeReportCodeCheckBreakingRecordsTask";
import writeReportCommentCoverageFunctionsTask from "./writeReportCommentCoverageFunctionsTask";
import writeReportCommitFileTagEvidencesTask from "./writeReportCommitFileTagEvidencesTask";
import writeReportCommitFileValuesTask from "./writeReportCommitFileValuesTask";
import writeReportCommitValuesTask from "./writeReportCommitValuesTask";
import writeReportDeveloperTestOfTimeFunction from './writeReportDeveloperTestOfTimeFunction';
import writeReportDuplicateFunctionsTask from "./writeReportDuplicateFunctionsTask";
import writeReportEmailMetricsTask from "./writeReportEmailMetricsTask";
import writeReportEmailModuleValuesTask from "./writeReportEmailModuleValuesTask";
import writeReportEmailValuesTask from "./writeReportEmailValuesTask";
import writeReportFunctionDependEmailsTask from "./writeReportFunctionDependEmailsTask";
import writeReportFunctionGraphTask from "./writeReportFunctionGraphTask";
import writeReportFunctionsTask from "./writeReportFunctionsTask";
import writeReportMetricTask from "./writeReportMetricTask";
import writeReportTask from "./writeReportTask";
import writeReportTestCoverageFunctionsTask from "./writeReportTestCoverageFunctionsTask";
import writeReportTopicEmailValuesTask from "./writeReportTopicEmailValuesTask";
import writeReportTopicFileValuesTask from "./writeReportTopicFileValuesTask";

export default [
  writeEmails,
  writeProjectCommitsTask,
  writeReportContributorBombSweeper,
  writeReportCodeCheckBreakingRecordsTask,
  writeReportCommentCoverageFunctionsTask,
  writeReportCommitFileTagEvidencesTask,
  writeReportCommitFileValuesTask,
  writeReportCommitValuesTask,
  writeReportDeveloperTestOfTimeFunction,
  writeReportDuplicateFunctionsTask,
  writeReportEmailMetricsTask,
  writeReportEmailModuleValuesTask,
  writeReportEmailValuesTask,
  writeReportFunctionDependEmailsTask,
  writeReportFunctionsTask,
  writeReportFunctionGraphTask,
  writeReportMetricTask,
  writeReportTask,
  writeReportTestCoverageFunctionsTask,
  writeReportTopicEmailValuesTask,
  writeReportTopicFileValuesTask,
];
