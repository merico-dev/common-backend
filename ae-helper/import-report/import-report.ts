import * as Fs from 'fs-extra';
import * as Path from 'path';
import { EntityManager } from 'typeorm';
import Protos from './protos';
import { mapFile, runTasks } from './tasks/utils';
import * as Minio from 'minio';
import Tmp from 'tmp-promise';
import Zlib from 'zlib';
import Tar from 'tar-fs';
import TASKS from './tasks';

export async function importFromDir(entityManager: EntityManager, projectId: string, reportId: string, protobufReportDir: string): Promise<void> {
  const mem = mapFile({protobufReportDir}, {filePath: 'report_of_all.pb'});
  const reportOfAll = Protos.ReportOfAll.decode(mem);

  await entityManager.transaction(async entityManager => {
    await runTasks({
      protobufReportDir,
      projectId,
      reportId,
      entityManager,
      reportOfAll,
    }, TASKS);
  });
}

export async function importFromBucket(entityManager: EntityManager, projectId: string, reportId: string, client: Minio.Client, bucketName: string, objectName: string): Promise<void> {
  const {path: tmpDir, cleanup} = await Tmp.dir({unsafeCleanup: true});
  try {
    const protobufReportFile = Path.join(tmpDir, 'pb_report.tar.gz');
    const protobufReportDir = Path.join(tmpDir, 'pb_report');
    await client.fGetObject(bucketName, objectName, protobufReportFile);
    await Fs.mkdir(protobufReportDir);
    await new Promise((resolve) => {
      Fs.createReadStream(protobufReportFile)
        .pipe(Zlib.createGunzip())
        .pipe(Tar.extract(protobufReportDir))
        .on('finish', () => {
          resolve()
        });
    });
    await importFromDir(entityManager, projectId, reportId, protobufReportDir);
  } finally {
    await cleanup();
  }
}
