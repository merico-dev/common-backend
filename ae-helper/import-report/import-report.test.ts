import * as Path from 'path';
import { importFromDir } from "./import-report";
import container from '../../ioc/inversify.config';
import * as COMMON_TYPES from '../../ioc/types';
import { Connection, Equal } from 'typeorm';
import { Email, Project, ProjectCommit, Report, ReportCodeCheckBreakingRecord, ReportCodeCheckBreakingRecordCode, ReportCommentCoverageFunction, ReportCommitFileTagEvidence, ReportCommitFileValue, ReportCommitValue, ReportDuplicateFunction, ReportEmailMetric, ReportEmailModuleValue, ReportEmailValue, ReportFunctionDependEmail, ReportFunctionGraph, ReportMetric, ReportTestCoverageFunction, ReportTopicEmailValue, ReportTopicFileValue } from '../../models';
import * as _ from 'lodash';

const projectId = '5024fc90-c18e-43aa-9c78-17b34c08f453';
const reportId = '1d7fa6f1-151a-4e36-8007-4406d0d80622';

beforeAll(async () => {
  const conn = await container.get<Promise<Connection>>(COMMON_TYPES.TYPEORM_CONNECTION_ASYNC);
  await conn.manager.delete(Project, projectId);
  await conn.manager.delete(Report, reportId);
  const project = new Project();
  project.id = projectId;
  project.project_name = 'gliderlabs/ssh';
  project.incomingReportId = reportId;
  project.git_url = 'https://github.com/gliderlabs/ssh.git';
  project.normalize_git_url = 'https://github.com/gliderlabs/ssh.git';
  project.readiness = 'UNDERWAY';
  await conn.manager.save(project);
});

test('import-report', async () => {
  const conn = await container.get<Promise<Connection>>(COMMON_TYPES.TYPEORM_CONNECTION_ASYNC);
  const manager = conn.manager;
  await importFromDir(manager, projectId, reportId, Path.join(__dirname, './test_protobuf'));
  expect(await manager.findOne(Email, {where: {email: Equal('progrium@gmail.com')}})).not.toBeNull();
  expect(await manager.count(ProjectCommit, {where: {project_id: Equal(projectId)}})).toMatchSnapshot();
  expect(await manager.count(ReportCodeCheckBreakingRecord, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportCodeCheckBreakingRecordCode, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportCommentCoverageFunction, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportCommitFileTagEvidence, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportCommitFileValue, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportCommitValue, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportDuplicateFunction, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportEmailMetric, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportEmailModuleValue, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportEmailValue, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportFunctionDependEmail, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportFunctionGraph, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportFunctionDependEmail, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportFunctionDependEmail, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportFunctionDependEmail, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.findOne(ReportMetric, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(_.omit(await manager.findOne(Report, {where: {report_id: Equal(reportId)}}), "create_time", "update_time")).toMatchSnapshot();
  expect(await manager.count(ReportTestCoverageFunction, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportTopicEmailValue, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
  expect(await manager.count(ReportTopicFileValue, {where: {report_id: Equal(reportId)}})).toMatchSnapshot();
});
