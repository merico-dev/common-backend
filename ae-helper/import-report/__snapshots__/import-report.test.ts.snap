// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import-report 1`] = `119`;

exports[`import-report 2`] = `106`;

exports[`import-report 3`] = `182`;

exports[`import-report 4`] = `196`;

exports[`import-report 5`] = `51`;

exports[`import-report 6`] = `304`;

exports[`import-report 7`] = `119`;

exports[`import-report 8`] = `4`;

exports[`import-report 9`] = `27`;

exports[`import-report 10`] = `72`;

exports[`import-report 11`] = `27`;

exports[`import-report 12`] = `311`;

exports[`import-report 13`] = `236`;

exports[`import-report 14`] = `311`;

exports[`import-report 15`] = `311`;

exports[`import-report 16`] = `311`;

exports[`import-report 17`] = `
ReportMetric {
  "ccg_snapshot_function_num": 196,
  "code_reusability": 0.979591846466064,
  "cyclomatic_total": 621,
  "doc_coverage": 25.5102043151855,
  "doc_coverage_function_num": 0,
  "doc_coverage_total_function_num": 196,
  "duplicate_function_num": 4,
  "git_tag_number": 9,
  "issue_blocker_number": 0,
  "issue_critical_number": 13,
  "issue_info_number": 11,
  "issue_major_number": 81,
  "issue_minor_number": 1,
  "issue_track_unique": true,
  "language_distribution": Object {
    "Dockerfile": 0.00070479407440871,
    "Go": 0.9992952346801758,
  },
  "magnetism": 0,
  "modularity": 75.3021850585938,
  "package_depend": 0,
  "popularity": null,
  "report_id": "1d7fa6f1-151a-4e36-8007-4406d0d80622",
  "robustness": 0.354941502889941,
  "static_test_coverage": 0.434426218271255,
  "static_test_coverage_function_num": 53,
  "static_test_coverage_total_function_num": 122,
  "velocity": 0,
}
`;

exports[`import-report 18`] = `
Object {
  "add_line": 4826,
  "commit_aspect_hash": "76cadaa318fd00fe3936af0243cb5daa9df84c73",
  "commit_dead_hash": "76cadaa318fd00fe3936af0243cb5daa9df84c73",
  "commit_dead_time": 2020-07-17T23:22:14.000Z,
  "commit_start_time": 2016-10-04T00:54:17.000Z,
  "default_ref": "refs/remotes/origin/master",
  "delete_line": 857,
  "detail_versions": Array [
    "report",
    "report_code_check_breaking_record",
    "report_code_check_breaking_record_code",
    "report_comment_coverage_function",
    "report_commit_file_tag_evidence",
    "report_commit_file_value",
    "report_commit_function",
    "report_commit_value",
    "report_duplicate_function",
    "report_duplicate_group",
    "report_email_metric",
    "report_email_module_value",
    "report_email_value",
    "report_function",
    "report_function_depend_email",
    "report_function_graph",
    "report_metric",
    "report_test_coverage_function",
    "report_topic_email_value",
    "report_topic_file_value",
  ],
  "dev_equivalent": 13292,
  "hasId": [Function],
  "id": "1d7fa6f1-151a-4e36-8007-4406d0d80622",
  "meta_dev_value": 0.000075233222991273,
  "num_commits": 119,
  "projectId": "5024fc90-c18e-43aa-9c78-17b34c08f453",
  "reload": [Function],
  "remove": [Function],
  "save": [Function],
  "version": "development",
}
`;

exports[`import-report 19`] = `122`;

exports[`import-report 20`] = `158`;

exports[`import-report 21`] = `86`;
